### PR TITLE
Conversations: fix badge count and automatic scrolling

### PR DESCRIPTION
**Pull request by svbergerem**
Fix for #4572, #4818 and #3998. I still need to add some tests but the code I wrote so far should be ready to review. Especially the controller changes might need some improvement. (was more or less trial and error)

### CODE REVIEW

File: app/controllers/conversations_controller.rb [hunk 1]
```diff
@@ -113,4 +116,11 @@ def contacts_data
         {value: contact_id, name: ERB::Util.h(Person.name_from_attrs(*name_attrs)) }
       }
   end
+
+  def set_read(conversation_id)
+    if @visibility = ConversationVisibility.where(:conversation_id => conversation_id, :person_id => current_user.person.id).first
+      @visibility.unread = 0
+      @visibility.save
+    end
+  end
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: Might as well move this method to the Conversation model.

File: app/views/conversations/index.haml [hunk 2]
```diff
@@ -17,7 +17,7 @@
         #conversation_inbox
           .stream.conversations
             - if @conversations.count > 0
-              = render :partial => 'conversations/conversation', :collection => @conversations, :locals => {:authors => @authors, :ordered_participants => @ordered_participants, :unread_counts => @unread_counts}
+              = render :partial => 'conversations/conversation', :collection => @conversations, :locals => {:authors => @authors, :ordered_participants => @ordered_participants, :unread_counts => @unread_counts, :selected_conversation_id => (@conversation ? @conversation.id : -1)}
```
Comment (jhass) [inline app/views/conversations/index.haml]: `@conversation.try(:id) || -1`, but I'm not sure how I feel about using -1 as a magic value here, why not check for nil in the target code?

File: app/views/conversations/_conversation.haml [hunk 3]
```diff
@@ -3,7 +3,9 @@
 -#   the COPYRIGHT file.
 
 .conversation-wrapper{ :"data-conversation-path" => conversation_path(conversation) }
-  .stream_element.conversation{:data=>{:guid=>conversation.id}, :class => ('unread' if unread_counts[conversation.id].to_i > 0)}
+  - conversation_class = (unread_counts[conversation.id].to_i > 0 ? 'unread' : '')
+  - conversation_class << (conversation.id == selected_conversation_id ? ' selected' : '')
```
Comment (jhass) [inline app/views/conversations/_conversation.haml]: Should probably be a helper method.

**Comment by Flaburgan** [comment]
Hm, it's probably not related to what you changed here but did you have a look at https://github.com/diaspora/diaspora/issues/5330 in the same time? This is a regression which is not in the current master, so it would be nice to see it fixed.

**Comment by svbergerem** [comment]
@Flaburgan No. As you said this is not related and in this PR I just want to fix the issues I already mentioned.

File: app/controllers/conversations_controller.rb (hunk 1 above)
Comment (svbergerem) [inline app/controllers/conversations_controller.rb]: done

File: app/views/conversations/index.haml (hunk 2 above)
Comment (svbergerem) [inline app/views/conversations/index.haml]: done

File: app/views/conversations/_conversation.haml (hunk 3 above)
Comment (svbergerem) [inline app/views/conversations/_conversation.haml]: done

**Comment by svbergerem** [comment]
All tests succeeded. Should be ready to review and merge.

**Comment by jhass** [comment]
Looks solid, thank you!

**Comment by Flaburgan** [comment]
Looks like we now have a problem trying to access the conversation page:

<pre>Processing by ConversationsController#index as HTML
  Rendered conversations/_conversation.haml (89.6ms)
  Rendered conversations/index.haml within layouts/with_header (92.1ms)
Completed 500 Internal Server Error in 353ms

ActionView::Template::Error (undefined method `conversation_class' for #<#<Class:0x0000326af217e0>:0x0000326afc3d10>):
    3: -#   the COPYRIGHT file.
    4: 
    5: .conversation-wrapper{ :"data-conversation-path" => conversation_path(conversation) }
    6:   .stream_element.conversation{:data=>{:guid=>conversation.id}, :class => conversation_class(conversation, unread_counts[conversation.id].to_i, selected_conversatio
n_id)}
    7:     .media
    8:       .img
    9:         - other_participants = ordered_participants[conversation.id] - [current_user.person]
  app/views/conversations/_conversation.haml:6:in `_app_views_conversations__conversation_haml__704969970413809694_108277623820'
  app/views/conversations/index.haml:20:in `_app_views_conversations_index_haml___1310477773073387960_108271293020'
  app/controllers/conversations_controller.rb:32:in `index'
  lib/rack/chrome_frame.rb:39:in `call'
  lib/unicorn_killer.rb:35:in `call'

**Comment by marienfressinaud** [comment]
@Flaburgan +1, but no problem on mobile

**Comment by Flaburgan** [comment]
Looks like something went wrong during the assets compilation, I cleaned them and re precompile them and now everything is fine.

**Pull request merged**